### PR TITLE
Bug fixed, exception and huge changes added!

### LINKED ISSUE #2: Decoding Suggestion Error
Hey 

When I try to run 'textsuggest', I get the following error

```
suggestion = suggestion.decode('utf-8')
S
/usr/bin/EnglishOpenWordList/S.txt
Traceback (most recent call last):
  File "/usr/bin/textsuggest", line 276, in <module>
    apply_suggestion(display_dialog_list(get_suggestions(current_word)))
  File "/usr/bin/textsuggest", line 208, in apply_suggestion
    suggestion = suggestion.decode('utf-8')
AttributeError: 'NoneType' object has no attribute 'decode'

```

I am running Kali 2016.1 Linux 64-bit on VMware

### PR DESCRIPTION
**Pull request by maateen**
**Fixed bugs:**
1. AttributeError: 'NoneType' object has no attribute 'decode'
2. FileNotFound Error

**Added changes:**
1. Layout & Language Dictionary
2. Bangla language support (Still working!)
3. Supporting multi-languages in get_suggestions function
4. Any phrase can be passed as parameter
5. Compatibility with Rofi, a drop-in replacement of dmenu (Still working!)

**Removed things:**
1. Unnecessary f.close(). When we're opening file using with then it automatically closes file after operation. Why do we need close() function, then?

Hope you will review, comment and merge if possible.

Sorry if I do any type of mistake. New to these things!

### CODE REVIEW

**Comment by bharadwaj-raju** [comment]
Great, but instead of having a full `(LanguageName)OpenWordList/`, can't the user just define them in `~/.Custom_Words.txt`?

Also, please remove the new Rofi changes you made — I've made a better Rofi configuration (dialog, colors, fonts, at-mouse-position) that is not yet committed.

**Comment by bharadwaj-raju** [comment]
Also, I'm just committing new, better Rofi support — after that is done, resolve conflicts again (sorry for bothering you with so many git conflicts).

**Comment by maateen** [comment]
Thanks for your comment. For other languages, using ~/.Custom_Words.txt won't be good. My languages has words over 10M. So, it's better to use a separate dictionary, isn't it?

Okay, commit your changes regarding Rofi. I will remove and resolve the conflicts and let you know so that you can merge my changes.

**Comment by bharadwaj-raju** [comment]
Suggestions: Additional dictionaries would be taken from `~/.TextSuggest_Dictionaries/` directory and if keyboard layout is not us/uk/English, we should take `(LanguageName)OpenWordList/` _and_ `EnglishOpenWordList/` both.

**Comment by bharadwaj-raju** [comment]
@maateen Switch to Rofi has been completed, look at above suggestion, then resolve conflicts.

**Comment by maateen** [comment]
@bharadwaj-raju 

Some update:

I have merged your latest commit and resolved current conflicts.

Suggestion: Let the dictionaries be in `OpenWordList/` in the parent directory, I mean, where the script is. In `OpenWordList/`directory, let make several directories with the language names like `English`, `Bangla`, `Hindi`, `Chinese` etc. In the specific language directory, we will keep the dictionaries.

If we use `~/.TextSuggest_Dictionaries/` then it won't be good. As many users delete their home folders by mistake. :smile: They will face issue after deleting.

> if keyboard layout is not us/uk/English, we should take (LanguageName)OpenWordList/ and EnglishOpenWordList/ both.

I agree with you on this point. 

Also I like to replace `Language_Layout.txt` with a function in a separate file like `Language_Layout.py` which will do the same thing. You know, file opening and reading isn't a fast method. Let's make the script more efficient. :smile:

**Comment by maateen** [comment]
@bharadwaj-raju 

Some Update:

I have replaced `Language_Layout.txt` with `languages.py` which contain a function `get_language_name()`. Also it has all languages and layouts definition. Let everyone update their own languages and layouts. :smile: 

I'm introducing `fonts.py` containing a function `get_font_name(language)` which will return a font name depending on the language User typing. Is it an excellent idea? :grinning: 

I have added **Bangla** support. Now please let take the decision about dictionaries.

File: fonts.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
+def get_font_name(langugae):
```
Comment (bharadwaj-raju) [inline fonts.py:1]: Spelling mistake!

**Comment by bharadwaj-raju** [comment]
> about dictionaries

Well, we can't include every language's dictionary in the repository, so we will provide English by default and tell users where they can find dictionaries for their language.

Also, in `fonts.py`, you have misspelled 'language', please correct.

But we are going to change the dictionary location to `~/.TextSuggest_Dictionaries`, since otherwise the packages install the dictionaries in `/usr/bin/` which is _not_ good. If a user deletes their home folder, they are going to have much bigger problems than just TextSuggest breaking :laughing:.

**Comment by bharadwaj-raju** [comment]
Just fix that, I'm ready to merge.

**Comment by bharadwaj-raju** [comment]
@maateen ^ See above.

**Comment by bharadwaj-raju** [comment]
I'll fix all that myself, merging!

File: fonts.py (hunk 1 above)
Comment (maateen) [inline fonts.py:1]: Feeling ashamed! :sob: I am fixing now.

**Comment by maateen** [comment]
I have fixed them. I will request another pull very soon.

> Well, we can't include every language's dictionary in the repository, so we will provide English by default and tell users where they can find dictionaries for their language.

I am feeling that it won't be great decision at all. Surely we can include them. There are many projects where all localized files are included with the project. Perhaps, I am separating my fork by naming `TextSuggestBangla`.

> But we are going to change the dictionary location to ~/.TextSuggest_Dictionaries, since otherwise the packages install the dictionaries in /usr/bin/ which is not good. If a user deletes their home folder, they are going to have much bigger problems than just TextSuggest breaking

Can all users on an OS get the benefit of TextSuggest then? Never. If we let them in /opt then all users can access them by our script. Please leave the idea. As there is already something called `Extra_Words.txt`.

**Comment by bharadwaj-raju** [comment]
> Perhaps, I am separating my fork by naming TextSuggestBangla

Let's stay together, please :smile:.

> There are many projects where all localized files are included with the project

Because they just include translations for a few words. This needs whole dictionaries. Now, there are _hundreds_ of languages. Our size would grow to 10s or hundreds of megabytes.

Can you look in `/usr/share/dict` if there are localized dictionaries there?

> If we let them in /opt

What about `/usr/share/dict`? That would be good, but needs root...

> Can all users on an OS

It isn't hard to copy a directory to another user's home. Many projects store configurations in `~`.

**Comment by bharadwaj-raju** [comment]
Most distributions include localization packs/spellcheck dictionaries (as used by LibreOffice and such). If we can find where they are, all our dictionary troubles would be solved.

EDIT: I seem to have found just the solution. Distros store dictionaries for every language the user has installed in `/usr/share/dict`. Mine:

``` bash
$ tree /usr/share/dict
/usr/share/dict
├── american-english
├── british-english
├── cracklib-small
├── README.select-wordlist
├── words -> /etc/dictionaries-common/words
└── words.pre-dictionaries-common -> american-english

0 directories, 6 files
```

**Comment by bharadwaj-raju** [comment]
@maateen What do you say? ^^

**Comment by bharadwaj-raju** [comment]
@maateen ^^ is that OK?

**Comment by maateen** [comment]
> Let's stay together, please!

Yah, I'm always with you. 

> Because they just include translations for a few words. This needs whole dictionaries. Now, there are hundreds of languages. Our size would grow to 10s or hundreds of megabytes.

Okay, let them be in home directory, I mean, in `~/.TextSuggest_Dictionaries`. And in `~/.TextSuggest_Dictionaries` the dictionary folder will be by name.

```
tree /usr/share/dict
/usr/share/dict
├── american-english
├── british-english
├── cracklib-small
├── README.select-wordlist
├── words -> /etc/dictionaries-common/words
└── words.pre-dictionaries-common -> american-english

0 directories, 6 files
```

**Comment by bharadwaj-raju** [comment]
Also, do you see anything like Bangla in the output of `ls /usr/share/dict/`?

**Comment by maateen** [comment]
Not at all. :( I making a Bangla dictionary by myself, porting an previous one from **IbusAvro.**

**Comment by maateen** [comment]
I implementing `~/.TextSuggest_Dictionaries` directory.

**Comment by bharadwaj-raju** [comment]
How about we add a `--language` option, so that user doesn't have to manually edit `languages.py` (the editing shall be done by script)?

Font can already be specified by `--font`.

**Comment by bharadwaj-raju** [comment]
I think this should be the new directory design:

```
$ tree ~/.config/textsuggest/
~/.config/textsuggest/
├── dictionaries/
|        ├── EnglishOpenWordList/
|        ├── Extra_Words.txt
|        └── LanguageOpenWordList/
├── Custom_Words.txt
├── textsuggest_history.txt
├── languages.py
├── fonts.py
```

I think we should remove `languages.py` (put it into main code or something) since specifying language with a `--language` option is much easier and cleaner than editing source code. Same for `fonts.py` (`--font` option)

**Comment by maateen** [comment]
```
$ tree ~/.config/textsuggest/
~/.config/textsuggest/
|── dictionaries/
|     |── English/
|     |── Bangla/
|     |-----Chinese/
|── Custom_Words.txt
|── textsuggest_history.txt
├── Extra_Words.txt
```

This one is a much mature design. But `languages.py` and `fonts.py` should be in the main script directory. I will follow it soon. :smile: `languages.py` should be there. None need edit the source code. We will do it. Let be everything automatic. Why should we use so much params? It bothers me a lot. :-1: The language should be automatically detected as same as I am doing.

**Comment by bharadwaj-raju** [comment]
Ah, I thought user has to edit languages.py to add a layout -> language association! :smile: We need to get a list of Linux keyboard layout codes and their languages. We'll keep languages.py

As for fonts.py... Remove it.

**Comment by maateen** [comment]
> Ah, I thought user has to edit languages.py to add a layout -> language association! :smile: We need to get a list of Linux keyboard layout codes and their languages. We'll keep languages.py

Thanks for the decision. Language like **Bangla** needs specific font like **SolaimanLipi**. Also **Chinese (Mandarin)** needs specific font. All language won't work with **Monospace**. So we need `fonts.py` too. When we will update `languages.py`, we will add fonts for the language too.

**Comment by bharadwaj-raju** [comment]
You don't need to worry, I'm doing all the redesign.

This is great, I am adding most to the languages dict: http://pastebin.com/v2vCPHjs

**Comment by bharadwaj-raju** [comment]
@maateen I think user if, say, speaks Chinese, should do `textsuggest --font MandarinFont 10`.

**Comment by bharadwaj-raju** [comment]
Instead of the `setxkbmap`, there is a better way: `echo $LANG`

**Comment by bharadwaj-raju** [comment]
Anyways, see:

``` python
languages = {
                    'bd'      : 'Bangla',
                    'us'      : 'English',
                    'uk'      : 'English',
                    'gb'      : 'English',
                    'ara'     : 'Arabic',
                    'cn'      : 'Chinese',
                    'de'      : 'German',
                    'jp'      : 'Japanese',
                    'ru'      : 'Russian',
                    'es'      : 'Spanish',
                    'se'      : 'Swedish',
                    'fi'      : 'Finnish',
                    'kr'      : 'Korean',
                    'pk'      : 'Urdu',
                    'fr'      : 'French',
                    'gr'      : 'Greek',
                    'tw'      : 'Chinese',
                    'ua'      : 'Ukrainian'
                }
```

**Comment by maateen** [comment]
> I think user if, say, speaks Chinese, should do textsuggest --font MandarinFont 10.

Nice work. :smile: But why shouldn't we make everything automatic? Let's keep the `fonts.py`. Please!

> Instead of the setxkbmap, there is a better way: echo $LANG

For me, it doesn't work well while `setxkbmap` gives me the different output even when I change the keyboard layout. We need it.

> ```
> languages = {
>                     'bd'      : 'Bangla',
>                     'us'      : 'English',
>                     'uk'      : 'English',
>                     'gb'      : 'English',
>                     'ara'     : 'Arabic',
>                     'cn'      : 'Chinese',
>                     'de'      : 'German',
>                     'jp'      : 'Japanese',
>                     'ru'      : 'Russian',
>                     'es'      : 'Spanish',
>                     'se'      : 'Swedish',
>                     'fi'      : 'Finnish',
>                     'kr'      : 'Korean',
>                     'pk'      : 'Urdu',
>                     'fr'      : 'French',
>                     'gr'      : 'Greek',
>                     'tw'      : 'Chinese',
>                     'ua'      : 'Ukrainian'
>                 }
> ```

Great work!

**Pull request merged**